### DISCUSSION TITLE
updating to version 3, Deployment is failng when some files don't exst

### DISCUSSION THREAD

**Discussion by takeaki-m** [serverless/serverless#12203] [Q&A] [answered]
I'm trying to migrage from serverless version 2 to 3.
In the process. I'm encoutering the problems. 
If anyone has a solution, please tell me some ideas.

### serverless.yml
My `serverless.yml`  is specifying resources file paths dynamitically based on `STAGE` variable.
The content of `A.yml` and `B.yml` are Cloudformation settings(e.g. codecommit, codebuild....)

#### version 2
```serverless.yml
service:
  name: sample_service
provider:
  name: aws
  stage: ${opt:stage, self:custom.defaultStage}
  region: ap-northeast-1
custom:
  defaultStage: test
resources:
  - ${file(./resources/${self:provider.stage}/A.yml)}
  - ${file(./resources/${self:provider.stage}/B.yml)}

```

#### version 3
```serverless.yml
frameworkVersion: '3'
service: sample_service
provider:
  name: aws
  stage: ${opt:stage, self:custom.defaultStage}
  region: ap-northeast-1
custom:
  defaultStage: test
resources:
  - ${file(./resources/${self:provider.stage}/A.yml)}
  - ${file(./resources/${self:provider.stage}/B.yml)}

```

### pacage structure

```
.
├── resources
│   ├── test
│   │   └── A.yml
│   │   └── B.yml
│   └── production
│   │   └── A.yml
└── serverless.yml
```

When I deploy serverless, I set stage based on stages

```bash
# test
sls deploy --stage test

# production
sls deploy --stage production
```

## RESULTS AND PROBLEMS
When I deploy `stage = production` , the results is different between v2 and v3. And Deployment of v3 failed.

### result version 2 
I got warining like below, but the deployment  succeded

```sh
A valid file to satisfy the declaration 'file(./resources/production/B.yml)' could not be found.
```

### result version 3
Deployment is failing with no detail messages.
So when I was debugging with `sls print` before `sls deploy`, the message is below

```sh


Error:
--
Cannot resolve serverless.yml: Variables resolution errored with:
Cannot resolve variable at "resources.10": Value not found at "file" source,

[Container]  Command did not exit successfully serverless print exit status 1
```

Any reference information like serverless option, package structure and so on would be appreciated.

**Comment by prince-biswaranjan** [comment]
https://www.serverless.com/framework/docs/guides/upgrading-v3#new-variable-resolver-engine

> Some edge cases (ambiguous configuration) now throw errors

Your error might be because of this change. In v2 it was just giving a warning and proceeding to deploy without considering that file.

**Reply by takeaki-m** [reply]
Thank you for showing the source. I haven't find the reason of the change!

**Comment by takeaki-m** [comment, marked as answer]
Finally, I have split serverless.yml file into each environment files.
In deploy process, renaming each environment files to `serverless.yml`

Here is one expamle.

package structure
```
.
├── resources
│   ├── test
│   │   └── A.yml
│   │   └── B.yml
│   └── production
│   │   └── A.yml
└── serverless
    └── production.yml
    └── staging.yml

```

In deploying, executing following commands

```
# variable stage is dynamically set to the environment name.
cp serverless/${stage}.yml serverless.yml

# refer to default serverless.yml file
sls deploy --stage ${stage}

```

Thank you for help in resolving the issue.

I'll be grad that my sample is helpful to the migration
